File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.py
class Solution(object):
    def hasAllCodes(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: bool
        """
        n = len(s)
        total = 1 << k
        
        # Quick impossible check
        if n < k or n - k + 1 < total:
            return False
        
        seen = [0] * total
        mask = 0
        needed = total
        limit = total - 1
        
        for i in range(n):
            # Build rolling k-bit integer
            mask = ((mask << 1) & limit) | (s[i] == '1')
            
            if i >= k - 1:
                if not seen[mask]:
                    seen[mask] = 1
                    needed -= 1
                    if needed == 0:
                        return True
        
        return False

File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.cpp
class Solution {
public:
    bool hasAllCodes(string s, int k) {
        int n = s.size();
        
        // Quick impossible case check
        if (n < k || n - k + 1 < (1 << k)) 
            return false;

        int total = 1 << k;
        vector<bool> seen(total, false);

        int mask = 0;
        int remaining = total;

        for (int i = 0; i < n; ++i) {
            // Build rolling k-bit number
            mask = ((mask << 1) & (total - 1)) | (s[i] - '0');

            if (i >= k - 1 && !seen[mask]) {
                seen[mask] = true;
                if (--remaining == 0)
                    return true;   // All codes found
            }
        }

        return false;
    }
};